### PR TITLE
[filedao] get recent blocks from staging buffer

### LINKED ISSUE #4489: failed to get block at height
### What version of iotex-core image (or code branch) are you using?
master

### What operating system and processor architecture are you using?
masOS

### What did you do? If possible, provide a recipe for reproducing the error.
query latest blocks from API

### What did you expect to see?
success

### What did you see instead?
{"level":"error","ts":"2024-11-14T01:25:51.074Z","msg":"failed to read block from dao","error":"failed to get block at height 33160559: key = 0000000000002732 doesn't exist: not exist in DB","errorVerbose":"not exist in DB\ngithub.com/iotexproject/iotex-core/db.init\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/db/builder.go:7\nruntime.doInit1\n\t/usr/local/go/src/runtime/proc.go:6740\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6707\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:249\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650\nkey = 0000000000002732 doesn't exist\ngithub.com/iotexproject/iotex-core/db.(*BoltDB).Get.func1\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/db/db_bolt.go:136\ngo.etcd.io/bbolt.(*DB).View\n\t/go/pkg/mod/go.etcd.io/bbolt@v1.3.7/db.go:893\ngithub.com/iotexproject/iotex-core/db.(*BoltDB).Get\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/db/db_bolt.go:129\ngithub.com/iotexproject/iotex-core/db.(*countingIndex).Get\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/db/counting_index.go:151\ngithub.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).getBlockStore\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/blockchain/filedao/filedao_v2_util.go:226\ngithub.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).getBlock\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/blockchain/filedao/filedao_v2_util.go:190\ngithub.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).GetBlockByHeight\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/blockchain/filedao/filedao_v2.go:188\ngithub.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAO).GetBlockByHeight\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/blockchain/filedao/filedao.go:205\ngithub.com/iotexproject/iotex-analyser/kernel.GetBlockByHeightFromBlockDAO\n\t/app/kernel/block.go:30\ngithub.com/iotexproject/iotex-analyser/server.(*runner).Start.func3\n\t/app/server/runner.go:238\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650\nfailed to get block at height 33160559\ngithub.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).GetBlockByHeight\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/blockchain/filedao/filedao_v2.go:190\ngithub.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAO).GetBlockByHeight\n\t/go/pkg/mod/github.com/iotexproject/iotex-core@v1.12.1-rc1.0.20240703182346-975014ac5905/blockchain/filedao/filedao.go:205\ngithub.com/iotexproject/iotex-analyser/kernel.GetBlockByHeightFromBlockDAO\n\t/app/kernel/block.go:30\ngithub.com/iotexproject/iotex-analyser/server.(*runner).Start.func3\n\t/app/server/runner.go:238\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650","pluginName":"erc20_meta","height":33160559,"tipHeight":33160559}

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
as title

Fixes #4489

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockchain/filedao/staging_buffer.go [hunk 1]
```diff
@@ -15,41 +17,53 @@ import (
 
 type (
 	stagingBuffer struct {
+		lock   sync.RWMutex
 		size   uint64
+		start  uint64
 		buffer []*block.Store
-		deser  *block.Deserializer
```
Comment (dustinxie) [inline blockchain/filedao/staging_buffer.go:20]: not needed

File: blockchain/filedao/filedao_v2_util.go [hunk 2]
```diff
@@ -151,11 +151,6 @@ func blockStoreKey(height uint64, header *FileHeader) uint64 {
 	return (height - header.Start) / header.BlockStoreSize
 }
 
-// stagingKey is the position of block in the staging buffer
-func stagingKey(height uint64, header *FileHeader) uint64 {
-	return (height - header.Start) % header.BlockStoreSize
-}
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2_util.go:157]: replaced by `stagingBuffer.slot()`

File: blockchain/filedao/filedao_v2_util.go [hunk 3]
```diff
@@ -215,12 +200,26 @@ func (fd *fileDAOv2) getReceipt(height uint64) ([]*action.Receipt, error) {
 	return fd.deser.ReceiptsFromBlockStoreProto(blockStore)
 }
 
+func (fd *fileDAOv2) getFromStagingBuffer(height uint64) (*block.Store, error) {
+	if fd.loadTip().Height-height >= fd.header.BlockStoreSize {
+		return nil, ErrNotSupported
+	}
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2_util.go:206]: staging buffer contains the top `BlockStoreSize` blocks, for example, if tip height = 100, it has blocks 100, 99, ... 85

File: blockchain/filedao/filedao_v2_util.go [hunk 4]
```diff
@@ -178,12 +173,7 @@ func (fd *fileDAOv2) getBlock(height uint64) (*block.Block, error) {
 		return nil, db.ErrNotExist
 	}
 	// check whether block in staging buffer or not
-	storeKey := blockStoreKey(height, fd.header)
-	if storeKey >= fd.blkStore.Size() {
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2_util.go:182]: root-cause of the bug:
the batch has not been written to DB yet, but `blkStore.Size()` already increased 1. This caused code go to L190 attempting to read the block from DB, which is not yet written to disk yet.

File: blockchain/filedao/staging_buffer.go [hunk 5]
```diff
@@ -15,41 +17,53 @@ import (
...
 }
 
+func (s *stagingBuffer) slot(height uint64) uint64 {
+	return (height - s.start) % s.size
+}
+
 func (s *stagingBuffer) Serialize() ([]byte, error) {
 	blkStores := []*iotextypes.BlockStore{}
 	// blob sidecar data are stored separately
+	s.lock.Lock()
```
Comment (envestcc) [inline blockchain/filedao/staging_buffer.go:62]: seems unnecessary, b/c `Put` and `Serialize` will not be called parallelly

File: blockchain/filedao/filedao_v2_util.go [hunk 6]
```diff
@@ -178,12 +173,7 @@ func (fd *fileDAOv2) getBlock(height uint64) (*block.Block, error) {
...
-		blkStore, err := fd.blkBuffer.Get(stagingKey(height, fd.header))
-		if err != nil {
-			return nil, err
-		}
+	if blkStore, err := fd.getFromStagingBuffer(height); err == nil {
```
Comment (envestcc) [inline blockchain/filedao/filedao_v2_util.go:176]: It is best to check for errors and throw if unexpected

File: blockchain/filedao/staging_buffer.go (hunk 5 above)
Comment (dustinxie) [reply blockchain/filedao/staging_buffer.go:62]: 1. in current code, they are not called parallelly. As 2 public methods, should consider synchronize access?
2. `Serialize` is actually also a read operation, so can change to `RLock`

File: blockchain/filedao/filedao_v2_util.go (hunk 3 above)
Comment (envestcc) [reply blockchain/filedao/filedao_v2_util.go:206]: The return value of `blkStore.Size()` is currently incorrect, and it seems that this PR did not fix it. Shouldn't the Size method be removed? If Size is used elsewhere, it should also be checked for similar issues.

Comment (dustinxie) [reply blockchain/filedao/filedao_v2_util.go]: will have another PR

File: blockchain/filedao/filedao_v2_util.go (hunk 6 above)
Comment (dustinxie) [reply blockchain/filedao/filedao_v2_util.go]: updated in 2nd commit

**Review by envestcc — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**